tobago-example/tobago-example-demo/src/test/java/org/apache/myfaces/tobago/example/demo/integration/FrontendBase.java

```diff
@@ -0,0 +1,156 @@
+        .filter(s -> s.endsWith(".test.js"))
+        .map(s -> s.substring("src/main/webapp/".length()))
+        .sorted()
+        .map(s -> s.substring(0, s.length() - 8) + ".xhtml")
+        .collect(Collectors.toList());
+  }
+
+  WebDriver getWebDriver(final String host, final Integer port) throws MalformedURLException {
+    if (firefoxDriver == null || ((RemoteWebDriver) firefoxDriver).getSessionId() == null) {
+      firefoxDriver = new RemoteWebDriver(new URL("http://" + host + ":" + port + "/wd/hub"), new FirefoxOptions());
+    }
+    return firefoxDriver;
+  }
+
+  List<WebElement> getJasmineResults(WebDriver webDriver, String url) {
+    final FluentWait<WebDriver> fluentWait = new FluentWait<>(webDriver)
+        .withTimeout(Duration.ofSeconds(60))
+        .pollingEvery(Duration.ofSeconds(1))
+        .ignoring(NoSuchElementException.class);
+    try {
+      fluentWait.until(driver -> driver.findElement(By.className("jasmine-overall-result")));
+    } catch (TimeoutException e) {
+      Assertions.fail(url + " timeout");
+    }
+
+    return webDriver.findElements(By.cssSelector(".jasmine-symbol-summary li"));
+  }
+
+  void parseJasmineResults(List<WebElement> results, String url) {
+    Assertions.assertTrue(results.size() > 0, url + " no results detected");
+
+    StringBuilder stringBuilder = new StringBuilder();
+    stringBuilder.append("\n");
+    stringBuilder.append(url);
+    for (WebElement result : results) {
+      stringBuilder.append("\n");
+      if ("jasmine-passed".equals(result.getAttribute("class"))) {
+        stringBuilder.append("✅ passed");
+      } else {
+        stringBuilder.append("❌ failed");
+      }
+      stringBuilder.append(": ");
+      stringBuilder.append(result.getAttribute("title"));
+    }
+    LOG.info(stringBuilder.toString());
+
+    for (WebElement result : results) {
+      Assertions.assertEquals("jasmine-passed", result.getAttribute("class"),
+          url + " " + result.getAttribute("title"));
+    }
+  }
+
+  String getTimeLeft(final LocalTime startTime, final int testSize, final int testNo) {
+    final LocalTime now = LocalTime.now();
+    final Duration completeWaitTime = Duration.between(startTime, now).dividedBy(testNo).multipliedBy(testSize);
+    final LocalTime endTime = LocalTime.from(startTime).plus(completeWaitTime);
+    final Duration timeLeft = Duration.between(LocalTime.now(), endTime);
+
+    if (timeLeft.toHours() > 0) {
+      return DurationFormatUtils.formatDuration(timeLeft.toMillis(), "H'h' m'm' s's'");
+    } else if (timeLeft.toMinutes() > 0) {
+      return DurationFormatUtils.formatDuration(timeLeft.toMillis(), "m'm' s's'");
+    } else if (timeLeft.toMillis() >= 0) {
+      return DurationFormatUtils.formatDuration(timeLeft.toMillis(), "s's'");
+    } else {
+      return "---";
+    }
+  }
+
+  int getFirefoxPort() {
+    return SELENIUM_FIREFOX.getFirstMappedPort();
+  }
+
+  int getTomcatPort() {
+    return TOMCAT.getFirstMappedPort();
+  }
```

Comment:
We don't need the selenium/firefox port and the tomcat port is fixed because we are calling the tomcat in the own docker network. We can call the tomcat with this url

```suggestion
  String getTomcatUrl() {
    return "http://" + TOMCAT_ALIAS + ":" + TOMCAT_PORT;
  }

```